modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php

```diff
@@ -77,36 +77,67 @@ public function enqueue_tracking_script() {
 	}
 
 	/**
-	 * On product lists or other non-product pages, add an event listener to "Add to Cart" button click
+	 * Render common default properties as a js args string.
 	 */
-	public function loop_session_events() {
+	public function get_common_properties() {
 		$blogid = Jetpack::get_option( 'id' );
 
-		// check for previous add-to-cart cart events
+		return "'blog_id': '" . esc_js( $blogid ) . "',
+			'ui': '" . esc_js( $this->get_user_id() ) . "',
+			'url': '" . esc_js( home_url() ) . "',
+			'woo_version': '" . esc_js( WC()->version ) . "',";
+	}
+
+	/**
+	 * Record an event with optional custom properties.
+	 *
+	 * @param string  $event_name The name of the event to record.
+	 * @param integer $product_id The id of the product relating to the event.
+	 * @param array   $properties Optional array of (key => value) event properties.
+	 */
+	public function record_event( $event_name, $product_id, $properties = array() ) {
+		$product = wc_get_product( $product_id );
+		if ( ! $product instanceof WC_Product ) {
+			return;
+		}
+		$product_details = $this->get_product_details( $product );
+
+		$extra_properties = '';
+		foreach ( $properties as $key => $value ) {
+			$extra_properties = $extra_properties . "'$key': '" . esc_js( $value ) . "', ";
+		}
+
+		wc_enqueue_js(
+			"_wca.push( {
+					'_en': '" . esc_js( $event_name ) . "',
+					'pi': '" . esc_js( $product_id ) . "',
+					'pn': '" . esc_js( $product_details['name'] ) . "',
+					'pc': '" . esc_js( $product_details['category'] ) . "',
+					'pp': '" . esc_js( $product_details['price'] ) . "',
+					'pt': '" . esc_js( $product_details['type'] ) . "'," .
+					$extra_properties . $this->get_common_properties() . '
+				} );'
+		);
+	}
+
+	/**
+	 * On product lists or other non-product pages, add an event listener to "Add to Cart" button click
+	 */
+	public function loop_session_events() {
+		// Check for previous events queued in session data.
 		if ( is_object( WC()->session ) ) {
```

Comment:
Should we check for `WC()->version` here as well, since we'll use it later?

Reply:
Similar to [previous comment](https://github.com/Automattic/jetpack/pull/15028#discussion_r394652146), I don't think this is necessary.We're dereferencing `WC()->session`, e.g. `WC()->session->get( 'wca_session_data' );`. With `version` we're just adding the value as a param, so no biggie if it's missing (and it's better to have the `woo_version` prop consistently available on events, even if the value is `NULL`).

Happy to put the check in if it's needed - i.e. let me know if there's something I'm missing here.

Reply:
Oh my bad, I missed your last comment!